lib/src/machine.rs

```diff
@@ -0,0 +1,379 @@
+use crate::{
+    ast::{Environment, Instruction, LangValue, Program, StackIdentifier},
+    error::RuntimeError,
+};
+use failure::Fallible;
+use std::iter;
+
+/// Wrapper around a vec of stacks, to make it a bit easier to initialize/use.
+///
+/// All stack manipulation logic should be implemented here, to keep it
+/// contained and scalable.
+/// This is zero-cost at runtime, see [newtype pattern](https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html).
+#[derive(Debug, PartialEq)]
+pub struct Stacks(Vec<Vec<LangValue>>);
+
+impl Stacks {
+    fn new(env: &Environment) -> Self {
+        Self(
+            iter::repeat(0)
+                .take(env.num_stacks)
+                .map(|_| Vec::new())
+                .collect(),
+        )
+    }
+
+    /// Gets a read-only view of the current set of stacks. Useful for
+    /// visualizations and the like.
+    pub fn get_all_stacks(&self) -> &[Vec<LangValue>] {
+        &self.0
+    }
+
+    /// Gets the stack with the given ID. If the stack doesn't exist, returns an
+    /// InvalidStackReference error.
+    fn get_stack(
+        &mut self,
+        stack_id: StackIdentifier,
+    ) -> Fallible<&mut Vec<LangValue>> {
+        self.0
+            .get_mut(stack_id)
+            .ok_or_else(|| RuntimeError::InvalidStackReference(stack_id).into())
+    }
+
+    /// Pushes the given value onto the given stack. The environment is needed
+    /// to compare the stack against a possible stack capacity rule. If the
+    /// stack reference is invalid or the stack is at capacity, an error is
+    /// returned.
+    fn push(
+        &mut self,
+        env: &Environment,
+        stack_id: StackIdentifier,
+        value: LangValue,
+    ) -> Fallible<()> {
+        let stack = self.get_stack(stack_id)?;
+
+        // If the stack is capacity, make sure we're not over it
+        if let Some(max_stack_size) = env.max_stack_size {
+            if stack.len() == max_stack_size {
+                return Err(RuntimeError::StackOverflow(stack_id).into());
```

Comment:
motion to make the error display this image https://i.imgur.com/Jmjv5To.jpg

Reply:
aww hell yeah, we can do that on the frontend